lib/src/conflicts.rs

```diff
@@ -743,21 +843,46 @@ pub async fn update_from_content(
     // copy.
     let mut old_content = Vec::with_capacity(content.len());
     let merge_hunk = extract_as_single_hunk(simplified_file_ids, store, path).await?;
-    materialize_merge_result(&merge_hunk, conflict_marker_style, &mut old_content).unwrap();
+    let materialized_marker_len =
+        materialize_merge_result(&merge_hunk, conflict_marker_style, &mut old_content).unwrap();
     if content == old_content {
         return Ok(file_ids.clone());
     }
 
+    let max_content_marker_len = find_max_conflict_marker_len(content);
+
+    // If all conflict markers in the file content are shorter than the materialized
+    // conflict markers, then the file is considered resolved.
+    if max_content_marker_len < materialized_marker_len.unwrap_or(MIN_CONFLICT_MARKER_LEN) {
+        let file_id = store.write_file(path, &mut &content[..]).await?;
+        return Ok(Merge::normal(file_id));
+    }
+
+    // Only accept the longest conflict markers in the file, also requiring that
+    // they are at least as long as the materialized conflict markers. We need to
+    // allow conflict markers to be longer than the materialized conflict markers,
+    // since the user could have removed the "fake" conflict markers from the file
+    // while editing, in which case their existing conflict markers will be
+    // considered longer than necessary the next time the file is snapshotted.
+    let expected_marker_len = materialized_marker_len
+        .unwrap_or_else(|| choose_conflict_marker_len(&merge_hunk))
+        .max(max_content_marker_len);
```

Comment:
Should we trust the marker length of the new `content`? Doesn't it mean any existing conflicts can be invalidated if user added long `<<<<<<<<<<<<<` literal by mistake?

Reply:
Yeah, they would be invalidated if the user adds a marker with length greater than or equal to the expected marker length. I'm not sure we can work around that though, since there will always be some length of marker that's invalid for them to add.

My goal was to make it so that if a file initially had fake markers of length N when it was materialized, then the user can always add/remove as many markers as they want with length <= N without breaking parsing, even if multiple snapshot operations happen during the process. I think it makes sense that adding markers of length > N could interfere with conflict resolution, since it makes it ambiguous whether the marker is a real one or not.